staging/src/k8s.io/apimachinery/pkg/util/runtime/runtime.go

```diff
@@ -161,3 +176,134 @@ func RecoverFromPanic(err *error) {
 	}
 }
+
+func newDedupingErrorHandler(cacheSize, errorDepth int, delta time.Duration) *dedupingErrorHandler {
+	d := &dedupingErrorHandler{
+		cache: lru.New(cacheSize),
+		count: make(map[errKey]errVal),
+
+		errorDepth: errorDepth,
+		delta:      delta,
+		clock:      clock.RealClock{},
+	}
+
+	d.cache.OnEvicted = func(key lru.Key, _ interface{}) {
+		// remove the associated entry in the count map when this key is evicted from the LRU cache
+		delete(d.count, key.(errKey))
+	}
+
+	return d
+}
+
+// dedupingErrorHandler provides a go routine safe error handler via handleErr.
+// It tracks errors via the caller stack, the error type and the error message (err.Error() value).
+// An error is considered unique based on these properties (see errKey).
+// To prevent from using an infinite amount of memory, it uses a LRU cache to purge old error values.
+type dedupingErrorHandler struct {
+	mutex sync.Mutex
+
+	// cache tracks (stack + type + message) and cleans up old entries in count as they roll off the cache
+	cache *lru.Cache
+	// count tracks (stack + type + message) -> (count + logged)
+	// since rudimentaryErrorBackoff rate limits HandleError to 1000 errors/second,
+	// this counter will effectively never overflow (nothing bad happens even if it does)
+	count map[errKey]errVal
+
+	// errorDepth is how many frames to skip from handleErr
+	errorDepth int
+
+	// delta is the minimum difference between the current time and
+	// errVal.logged required for us to log the associated error again
+	delta time.Duration
+
+	// clock allows us to control time in unit tests
+	// TODO add unit tests
+	clock clock.Clock
+}
+
+// errKey tracks uniqueness based on the caller's stack and the type/message of the error
+// it is stored in both the count map and the LRU cache
+// it is removed from the count map when it gets evicted from the LRU cache
+// it is comparable via ==
+type errKey struct {
+	stack   string
+	errType reflect.Type
+	// message is the err.Error() value
+	message string
+}
+
+// errVal tracks how many times we have seen an error, and the last time we logged it
+type errVal struct {
+	count  uint64
+	logged time.Time
+}
+
+// handleErr logs the given error if it is considered new or "not recently logged"
+// currently it logs errors whenever:
+// 1. the associated errKey does not exist in d.count
+// 2. the associated counter is a power of two
+// 3. the associated logged time's delta from the current time is greater than d.delta
+func (d *dedupingErrorHandler) handleErr(err error) {
+	d.mutex.Lock()
+	defer d.mutex.Unlock()
+
+	// we must determine our stack in this function since getStack counts frames
+	stack := d.getStack()
+	key := errKey{stack: stack, errType: reflect.TypeOf(err), message: err.Error()}
+
+	// increment our counter
+	val, isOldErr := d.count[key]
+	isNewErr := !isOldErr
+	val.count++
+
+	if isNewErr {
+		// we did not find the error, so add
+		// the associated entry in the LRU cache
+		d.cache.Add(key, nil)
```

Comment:
So as I read this code, I come to the conclusion that you use the LRU to auto-roll items out of the cache, but a few lines up you will probably decide you need to fuzzy match a particular key based on message, so you want easy access to the keys.

If I'm reading it write, you need to place a comment in the exact spot you want to do the fuzzy key matching logic.

Reply:
I see your point though - as it stands (without fuzzy matching), I could drop the map entirely and use only the LRU cache.

Reply:
> I originally wrote this with fuzzy matching via a Levenshtein ratio master...enj:enj/i/handle_error_dedupe_bu0
> 
> I found it to be really complex so I dropped it for a simpler direct comparison.
> 
> I suppose I could do a simpler approach that involves iterating over the entire map. O(1000) seems OK.

https://github.com/charlesvdv/fuzmatch with the simple ratio and a reasonable threshold?

Reply:
> I originally wrote this with fuzzy matching via a Levenshtein ratio master...enj:enj/i/handle_error_dedupe_bu0
> 
> I found it to be really complex so I dropped it for a simpler direct comparison.
> 
> I suppose I could do a simpler approach that involves iterating over the entire map. O(1000) seems OK.
> https://github.com/charlesvdv/fuzmatch with the simple ratio and a reasonable threshold?

On the message.  The stack has to match exactly.